### PR TITLE
ensure we finish setting up background session before providing backg…

### PR DESCRIPTION
**Pull request by thuyBox**
…round session tasks

### CODE REVIEW

**Comment by boxcla** [comment]
Verified that @thuyBox has signed the CLA. Thanks for the pull request!

File: BoxContentSDK/BoxContentSDK/BOXURLSessionManager.m [hunk 1]
```diff
@@ -57,7 +57,10 @@ - (id)initWithBackgroundSessionId:(NSString *)backgroundSessionId sessionTask:(N
 
 @end
 
-@interface BOXURLSessionManager() <NSURLSessionDelegate, NSURLSessionTaskDelegate, NSURLSessionDataDelegate, NSURLSessionDownloadDelegate, NSURLSessionStreamDelegate>
+@interface BOXURLSessionManager() <NSURLSessionDelegate, NSURLSessionTaskDelegate, NSURLSessionDataDelegate, NSURLSessionDownloadDelegate, NSURLSessionStreamDelegate> {
```
Comment (boxtcarpel) [inline BoxContentSDK/BoxContentSDK/BOXURLSessionManager.m:60]: Don't need SessionDelegate and SessionTaskDelegate, the others are extensions of those so they are already included.

File: BoxContentSDK/BoxContentSDK/BOXURLSessionManager.m [hunk 2]
```diff
@@ -83,10 +86,17 @@ @interface BOXURLSessionManager() <NSURLSessionDelegate, NSURLSessionTaskDelegat
...
+//A map to associate backgroundSessionId and equivalent background NSURLSession objects that this BOXURLSessionManager knows about
+//If inside an app, there will be one background session for app, and 0 or more from previous runs of extensions
+//If inside an extension, there will be one background session for extension
 @property (nonatomic, readonly, strong) NSMutableDictionary<NSString *, NSURLSession *> *backgroundSessionIdToSession;
 
+//A cache client to cache background session task-related data to file system for retrieval after app restarts
 @property (nonatomic, readwrite, strong) BOXURLSessionCacheClient *cacheClient;
 
+//Indicate if background session's setup has completed
+@property (nonatomic, readwrite, assign) BOOL finishSettingUpBackgroundSession;
```
Comment (boxtcarpel) [inline BoxContentSDK/BoxContentSDK/BOXURLSessionManager.m:98]: Minor: Finish reads like a verb here, should this be something like 'didFinishSettingUp...'?

File: BoxContentSDK/BoxContentSDK/BOXURLSessionManager.m [hunk 3]
```diff
@@ -57,7 +57,10 @@ - (id)initWithBackgroundSessionId:(NSString *)backgroundSessionId sessionTask:(N
 
 @end
 
-@interface BOXURLSessionManager() <NSURLSessionDelegate, NSURLSessionTaskDelegate, NSURLSessionDataDelegate, NSURLSessionDownloadDelegate, NSURLSessionStreamDelegate>
+@interface BOXURLSessionManager() <NSURLSessionDelegate, NSURLSessionTaskDelegate, NSURLSessionDataDelegate, NSURLSessionDownloadDelegate, NSURLSessionStreamDelegate> {
+    //A semaphore to make sure we wait until background session finishes setting up before providing background session tasks
+    dispatch_semaphore_t backgroundSessionSetupSemphore;
```
Comment (boxtcarpel) [inline BoxContentSDK/BoxContentSDK/BOXURLSessionManager.m:62]: name it with an underscore so it reads like an instance variable?

**Review by boxtcarpel — APPROVED** [review]
Have we been able to measure the overhead?

File: BoxContentSDK/BoxContentSDK/BOXURLSessionManager.m [hunk 4]
```diff
@@ -115,10 +125,19 @@ - (id)init
         _backgroundSessionIdToSession = [NSMutableDictionary new];
         [self createDefaultSession];
         [self createProgressSession];
+        self.finishSettingUpBackgroundSession = NO;
```
Comment (boxtcarpel) [inline BoxContentSDK/BoxContentSDK/BOXURLSessionManager.m:128]: Minor: Use the ivar directly

**Review by boxtcarpel — APPROVED** [review]

**Pull request merged**